pkg/project/auth/watch.go

```diff
@@ -0,0 +1,196 @@
+package auth
+
+import (
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/golang/glog"
+
+	kapi "k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/auth/user"
+	"k8s.io/kubernetes/pkg/util"
+	"k8s.io/kubernetes/pkg/util/sets"
+	"k8s.io/kubernetes/pkg/watch"
+
+	projectapi "github.com/openshift/origin/pkg/project/api"
+	projectcache "github.com/openshift/origin/pkg/project/cache"
+)
+
+type CacheWatcher interface {
+	GroupMembershipChanged(namespaceName string, removedUsers, removedGroups, addedUsers, addedGroups sets.String)
+}
+
+// userProjectWatcher converts a native etcd watch to a watch.Interface.
+type userProjectWatcher struct {
+	username string
+	groups   []string
+
+	cacheIncoming  chan watch.Event
+	cacheError     chan error
+	cacheCallEnded chan struct{}
+
+	outgoing chan watch.Event
+	userStop chan struct{}
+	stopped  bool
+	stopLock sync.Mutex
+
+	// Injectable for testing. Send the event down the outgoing channel.
+	emit func(watch.Event)
+
+	projectCache *projectcache.ProjectCache
+	authCache    *AuthorizationCache
+
+	knownProjects sets.String
+}
+
+var (
+	watchChannelHWM util.HighWaterMark
+)
+
+// watchWaitDuration is the amount of time to wait for an error from watch.
+const watchWaitDuration = 100 * time.Millisecond
+
+func NewUserProjectWatcher(username string, groups []string, projectCache *projectcache.ProjectCache, authCache *AuthorizationCache) *userProjectWatcher {
+	userInfo := &user.DefaultInfo{Name: username, Groups: groups}
+	namespaces, _ := authCache.List(userInfo)
+	namespaceNames := sets.String{}
+	for _, namespace := range namespaces.Items {
+		namespaceNames.Insert(namespace.Name)
+	}
+
+	w := &userProjectWatcher{
+		username: username,
+		groups:   groups,
+
+		cacheIncoming: make(chan watch.Event, 100),
+		cacheError:    make(chan error, 1),
+		outgoing:      make(chan watch.Event),
+		userStop:      make(chan struct{}),
+
+		projectCache:  projectCache,
+		authCache:     authCache,
+		knownProjects: namespaceNames,
+	}
+	w.emit = func(e watch.Event) {
+		w.outgoing <- e
+	}
+	return w
+}
+
+func (w *userProjectWatcher) GroupMembershipChanged(namespaceName string, removedUsers, removedGroups, addedUsers, addedGroups sets.String) {
+	removed := removedUsers.Has(w.username) || removedGroups.HasAny(w.groups...)
+	added := addedUsers.Has(w.username) || addedGroups.HasAny(w.groups...)
+
+	switch {
+	// do nothing if we're net neutral
+	case removed && added:
+	case !removed && !added:
+
+	case removed:
+		if !w.knownProjects.Has(namespaceName) {
+			return
+		}
+		w.knownProjects.Delete(namespaceName)
+		select {
+		case w.cacheIncoming <- watch.Event{
+			Type:   watch.Deleted,
+			Object: convertNamespace(&kapi.Namespace{ObjectMeta: kapi.ObjectMeta{Name: namespaceName}}),
+		}:
+
+		case <-time.After(10 * time.Millisecond):
```

Comment:
Today project metadata updates are rare, just display name / description,
not common.  So yes this is useful as is with the current state of the
world.  It will at least make the delete project from the console look
better since that project will actually disappear instead of sitting there
with a terminating icon.

In the future if we start allowing users to do more with their projects, by
whitelisting an openshift.io-console (or whatever) annotation namespace
(its been discussed), then maybe updates would be helpful.

On Wednesday, January 27, 2016, David Eads <notifications@github.com> wrote:

> In pkg/project/auth/watch.go
> <https://github.com/openshift/origin/pull/6828#discussion_r50999524>:
>
> > +	// do nothing if we're net neutral
> > +	case removed && added:
> > +	case !removed && !added:
> > +
> > +	case removed:
> > +		if !w.knownProjects.Has(namespaceName) {
> > +			return
> > +		}
> > +		w.knownProjects.Delete(namespaceName)
> > +		select {
> > +		case w.cacheIncoming <- watch.Event{
> > +			Type:   watch.Deleted,
> > +			Object: convertNamespace(&kapi.Namespace{ObjectMeta: kapi.ObjectMeta{Name: namespaceName}}),
> > +		}:
> > +
> > +		case <-time.After(10 * time.Millisecond):
>
> Stef wants watch, but doesn't need it. I think we want watch too, but
> don't need it.
>
> Do we *want* it with the restrictions I laid out? It's pretty simple to
> take this to completion as long as I don't try to handle coalescing of
> namespace updates with coherent ordering to the auth cache.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/6828/files#r50999524>.
>